Visit the transforms before handling delayed assattr nodes

marmarek opened the pull request:
<!--
Thank you for submitting a PR to astroid!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Write a good description on what the PR does.
- [ ] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

When importing Gtk, it looks like this:

    import gi
    gi.require_version('Gtk', '3.0')
    from gi.repository import Gtk

It is vital that gi.require_version() is made before related 'from gi.repository import ...'. The brain_gi tries to do that using transforms. And it works unless Gtk is imported as part of delayed assattr handling.

Fix this by handling transforms earlier.

Closes https://github.com/pylint-dev/astroid/issues/2190
Closes https://github.com/pylint-dev/pylint/issues/6352

marmarek committed:
Add early transforms before handling delayed assattr nodes

When importing Gtk, it looks like this:

    import gi
    gi.require_version('Gtk', '3.0')
    from gi.repository import Gtk

It is vital that gi.require_version() is made before related 'from
gi.repository import ...'. The brain_gi tries to do that using
transforms. And it works unless Gtk is imported as part of delayed
assattr handling.

Fix this by adding early transforms that are called before delayed
assattr.

Fixes https://github.com/pylint-dev/astroid/issues/2190
Fixes https://github.com/pylint-dev/pylint/issues/6352

marmarek commented:
My change made one test fail now:
```
____________________ TypingBrain.test_typing_generic_slots _____________________

self = <tests.brain.test_brain.TypingBrain testMethod=test_typing_generic_slots>

    def test_typing_generic_slots(self):
        """Test slots for Generic subclass."""
        node = builder.extract_node(
            """
        from typing import Generic, TypeVar
        T = TypeVar('T')
        class A(Generic[T]):
            __slots__ = ['value']
            def __init__(self, value):
                self.value = value
        """
        )
        inferred = next(node.infer())
        slots = inferred.slots()
>       assert len(slots) == 1
E       TypeError: object of type 'NoneType' has no len()
```

I don't fully understand why, yet, but I think one option to make my change less invasive is to add "early transforms" and leave the original transforms in the old place. And then make brain_gi use the early transforms for the gi.require_version thing. Is that okay?

Pierre-Sassoulas commented:
Thank you for the MR. I agree with the proposed early transform, it feels like a big API decision but applying the transform twice so it's possible to apply transform after delayed attr creation is a lot worse. Let's hear another maintainer opinion before committing to it, it's not like I'm a brain expert myself.

DanielNoord commented:
Have you also tried running the `pylint` testsuite with this version of `astroid`? I'm wondering if this is a fix we can indeed make, the failing test doesn't make me too hopeful.

marmarek commented:
When running pylint test suite on plain main branch I got 12 failures already, but my change introduces 13th: `test_functional[assigning_non_slot_4509]` (`assigning-non-slot` message is not raised).
The failure here is also related to slots, so it's probably the same thing.

I'll change my patch to not affect other brains.

marmarek committed:
Add a test for brain_gi

Just a simple one, that checks if the correct Gtk version gets
installed.

Note this test is checks for the linked regression only if both Gtk3 and
Gtk4 are installed. Otherwise, it checks if brain_gi works at all.

marmarek commented:
While at it, I've added also some basic test for brain_gi.

marmarek commented:
> tests\brain\test_gi.py s                                                 [ 20%]

So, the test I added doesn't run in CI (which also explain minimal test coverage of brain_gi.py). Is it worth installing "gi" python module + Gtk3 (and maybe Gtk4 too?) in there to make the test run? Or is it enough for it to run when called manually locally (assuming those dependencies are installed)?

DanielNoord commented:
Yeah I don't think we want to add those dependencies to the CI.

I do wonder if this warrants the API change. I can't really comment on that part (and I think most maintainers are a bit afraid of making such a decision). @Pierre-Sassoulas @jacobtylerwalls Do you have any opinions here?

jacobtylerwalls commented:
> I do wonder if this warrants the API change.

Yeah, I'm reluctant to add an early transforms API for this. We should consider alternatives:
- in _register_require_version(): manipulate sys.modules or use importlib.reload to reimport gi if necessary
- remove brain_gi in favor of a new plugin to be homed at pylint-dev/pylint-gi

DanielNoord commented:
A new plugin is probably what we would want considering the maintainer time we have for both `astroid` and `pylint`. Closing this for now, sorry!
